### PR TITLE
Introduce Mapping ActionListener

### PR DESCRIPTION
**Pull request by original-brownbear**
* The motivation here is to make reasoning about chains of `ActionListener` a little easier, by providing a safe method for nesting `ActionListener` that guarantees that a response is never dropped. Also, it dries up the code a little by removing the need to repeat `listener::onFailure` and `listener.onResponse` over and over.
* Refactored a number of obvious/easy spots to use the new listener constructor

### CODE REVIEW

**Comment by original-brownbear** [comment]
Motivated in part by #39504 where this kind of wrapper was nice to safely move the bulk execution path to async :)

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java [hunk 1]
```diff
@@ -815,10 +815,10 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
                     getRequest = new GetRequest(indexedShapeIndex, indexedShapeType, indexedShapeId);
                 }
                 getRequest.routing(indexedShapeRouting);
-                fetch(client, getRequest, indexedShapePath, ActionListener.wrap(builder-> {
+                fetch(client, getRequest, indexedShapePath, ActionListener.map(listener, builder -> {
                     supplier.set(builder);
-                    listener.onResponse(null);
-                }, listener::onFailure));
+                    return null;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: With side-effect, I think ActionListener.wrap is more suitable than the new method.

**Review by dnhatn — APPROVED** [review]
I am +1 to this change for it's safer and less verbose than `ActionListener.wrap`. I left two minor comments.

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 2]
```diff
@@ -83,6 +84,20 @@ public void onFailure(Exception e) {
         return wrap(r -> runnable.run(), e -> runnable.run());
     }
 
+    /**
+     * Creates a listener that wraps another listener, mapping response values via the given mapping function and passing along
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/ActionListener.java]: Makes sense to me :) Renamed the parameter and slightly reworded the Javadoc.

**Comment by original-brownbear** [comment]
Thanks @dnhatn !

**Comment by ywelsch** [comment]
I wonder if this would all be nicer to write by using `default` methods in ActionListener instead of `static` ones. 

This would allow us to write `listener.map(CreateSnapshotResponse::new)` instead of `ActionListener.map(listener, CreateSnapshotResponse::new)`.

**Comment by original-brownbear** [comment]
@ywelsch I thought the same, but then I figured you can't make `default` methods `final` -> I felt like it was less readable unless you knew that no one would ever override `map`?

**Pull request merged**